### ISSUE TITLE
DomainFree lambdas and invalid identifiers might appear after the type checking of a term

### ISSUE THREAD

**Issue by francoisthire** [Deducteam/Dedukti#51]
The function `solve` in module `matching` introduces domain free lambdas and invalid identifiers. This function is called when the module `Reduction` has to solve a Miller's pattern.

I have a complicated example (that comes from the arithmetic library) where the inference of a term fails because Dedukti cannot guess the type of a domain free lambda. However, there is no such lambda in the original file: This lambda has been introduced by the matching function.

This is also problematic since it might happen that this terms, once printed using the module `Pp` cannot be parsed again since this function introduces invalid identifiers.

While I think the second problem can be fixed easily, for example by changing the identifier, the first problem might be more complicated to fix since no type information is carried during the reduction. However, for the moment, I don't have a nice and simple example that reveal that behavior.

**Comment by francoisthire** [comment]
Here, such an example:

```
nat : Type.

0 : nat.

S : nat -> nat.

def A : (nat -> nat) -> (nat -> nat) -> nat -> nat.

[f] A (x => f x) f --> f.

#EVAL A (x : nat => 0) (x : nat => 0).
```

**Comment by francoisthire** [comment]
Fixed in #60
